atom/src/catom.cpp

```diff
@@ -370,6 +372,88 @@ CAtom_sizeof( CAtom* self, PyObject* args )
     return PyLong_FromSsize_t( size );
 }
 
+PyObject*
+CAtom_getstate( CAtom* self )
+{
+    cppy::ptr stateptr = PyDict_New();
+    if ( !stateptr )
+        return PyErr_NoMemory();  // LCOV_EXCL_LINE
+
+    cppy::ptr selfptr(pyobject_cast(self), true);
+
+    // Copy __dict__ if present. Using hasattr / getattr makes it slower
+    // than the py version hence the _PyObject_GetDictPtr.
+    if ( PyObject** dict = _PyObject_GetDictPtr( selfptr.get() ) )
+    {
+        if ( PyDict_Update( stateptr.get(), dict[0] ) )
+            return 0;
+    }
+
+    // Copy __slots__ if present. This assumes copyreg._slotnames was called
+    // during AtomMeta's initialization
+    {
+        cppy::ptr typeptr = PyObject_Type(selfptr.get());
+        if (!typeptr)
+            return 0;
+        cppy::ptr slotnamesptr = typeptr.getattr("__slotnames__");
+        if (!slotnamesptr.get())
+            return 0;
+        if (!PyList_CheckExact(slotnamesptr.get()))
+            return cppy::system_error( "slot names" );
+        for ( Py_ssize_t i=0; i < PyList_GET_SIZE(slotnamesptr.get()); i++ )
+        {
+            PyObject *name = PyList_GET_ITEM(slotnamesptr.get(), i);
+            cppy::ptr value = selfptr.getattr(name);
+            if (!value || PyDict_SetItem(stateptr.get(), name, value.get()) )
+                return  0;
+        }
+    }
+
+    cppy::ptr membersptr = selfptr.getattr("__atom_members__");
+    if ( !membersptr || !PyDict_CheckExact( membersptr.get() ) )
+        return cppy::system_error( "atom members" );
+
+    PyObject *name, *member;
+    Py_ssize_t pos = 0;
+    while ( PyDict_Next(membersptr.get(), &pos, &name, &member) ) {
+        cppy::ptr value = selfptr.getattr(name);
+        if (!value || PyDict_SetItem(stateptr.get(), name, value.get()) )
+            return  0;
+    }
```

Comment:
Here since we have a direct access to the member we could check the setattr mode and exclude member with a constant mode, signal mode, property mode and a None context which we are sure are read only. 
Alternatively we could add an attribute on the member signaling if it can be pickled. This would allow to keep using the C getstate with custom members. WDYT ?